File: play-with-lambda-expressions/play_with_lambda/adapters/readline_ui.py
"""Terminal-based REPL interface with readline support."""

from __future__ import annotations


class ReadlineUI:
    """Interactive terminal UI with history and tab completion.

    Implements the UserInterface protocol for real terminal interaction.
    """

    _COMMANDS: list[str] = [
        ":reduce", ":step", ":trace",
        ":free", ":bound", ":binding",
        ":random", ":let", ":list",
        ":help", ":quit",
    ]

    def __init__(self) -> None:
        try:
            import readline

            def completer(text: str, state: int) -> str | None:
                matches = [c for c in self._COMMANDS if c.startswith(text)]
                return matches[state] if state < len(matches) else None

            readline.set_completer(completer)
            readline.parse_and_bind("tab: complete")
        except ImportError:
            pass

    def read_input(self) -> str | None:
        """Read one line from the user.

        Returns:
            The input string, or None on EOF / KeyboardInterrupt.
        """
        try:
            return input("λ> ")
        except (EOFError, KeyboardInterrupt):
            print()
            return None

    def display(self, text: str) -> None:
        """Print output to stdout."""
        print(text)

    def display_error(self, text: str) -> None:
        """Print an error message to stdout."""
        print(f"Error: {text}")

File: play-with-lambda-expressions/src/lambda_calculus/src/adapters/readline_ui.rs
//! Terminal-based REPL interface with rustyline support.

use rustyline::error::ReadlineError;
use rustyline::DefaultEditor;
use crate::ports::UserInterface;

/// Interactive terminal UI with history and tab completion.
pub struct ReadlineUI {
    editor: DefaultEditor,
}

impl ReadlineUI {
    pub fn new() -> Self {
        let editor = DefaultEditor::new().expect("failed to create readline editor");
        Self { editor }
    }
}

impl UserInterface for ReadlineUI {
    fn read_input(&mut self) -> Option<String> {
        match self.editor.readline("λ> ") {
            Ok(line) => {
                let _ = self.editor.add_history_entry(&line);
                Some(line)
            }
            Err(ReadlineError::Interrupted | ReadlineError::Eof) => {
                println!();
                None
            }
            Err(_) => None,
        }
    }

    fn display(&mut self, text: &str) {
        println!("{}", text);
    }

    fn display_error(&mut self, text: &str) {
        println!("Error: {}", text);
    }
}
